src/util/compat.js

```diff
@@ -0,0 +1,66 @@
+// @flow
+
+export type Compatible = {
+  +__compat__: {|
+    +type: string,
+    +version: string,
+  |},
+};
+
+export function toCompat<T>(
+  type: string,
+  version: string,
+  obj: T
+): T & Compatible {
+  if (typeof obj !== "object") {
+    throw new Error(
+      `Tried to compatibilize ${typeof obj}, but only objects are supported`
+    );
+  }
+  if (obj == null) {
+    throw new Error(`Tried to compatibilize null value`);
+  }
+  if (obj.__compat__ != null) {
```

Comment:
Would prefer `"__compat__" in obj` here: an explicitly set `null` or `undefined` anywhere on the proto should be sufficient to fail.